### DISCUSSION TITLE
How to iterate over and unsubscribe from all topics

### DISCUSSION THREAD

**Discussion by thorlancaster** [uNetworking/uWebSockets#1391] [Q&A] [answered]
I am making a uWebSocket application where I have to be able to unsubscribe users from all subscribed topics. I am trying to do this with a lambda function (?) that iterates over all of a socket's subscribed topics, puts them into a std::vector (to avoid concurrent modification), and then unsubscribes from each one.

Here's what I have so far:
`void unsubscribeAll(uWS::WebSocket<true, true, PerSocketData> *ws){
        std::vector<std::string> topics;

        ws->iterateTopics([](std::string topic){
            topics.push_back(topic);
        });

        for(int x = 0; x < topics.size(); x++){
            ws->unsubscribe(topics[x]);
        }
    }`

Unfortunately, it's giving me an error: `No viable conversion from '(lambda at /home/thor/lsd/Synchronizr.h:281:27)' to 'MoveOnlyFunction<void (std::string_view)>' (aka 'any_invocable<void (std::basic_string_view<char>)>')`

How do I make the above code work? Currently I am storing the subscriptions in the userData, but that isn't optimal.

**Comment by ghost** [comment, marked as answer]
A university student working on project /home/thor/**lsd**/ brings me back 10 years, but I preferred mushrooms 🍄 😜 

In any case - it's string_view not string

**Reply by thorlancaster** [reply]
@alexhultman Thank you for pointing that out. As a longtime Java programmer who has just started C++, I have a hard time following some of the errors, even with the help of CLion.

I managed to figure it out:

```
void unsubscribeAll(uWS::WebSocket<true, true, PerSocketData> *ws){
    std::vector<std::string> topics;
    ws->iterateTopics([&topics](std::string_view topic){
        std::string str = {topic.data(), topic.size()};
        topics.push_back(str);
    });
    for(const std::string& topic : topics){
        ws->unsubscribe(topic);
    }
}
```
Thank you again for your awesome library. It's already miles better than the Java WebSocket library that I used last year which randomly hung for a minute or so at the worst times.

P.S. LSD stands for Live Synchronizr Daemon. It's the server for a live sports stats feed and will be on my Github once it's feature-complete. Admittedly I have thought of :seedling: a few times when my C++ was misbehaving :laughing:
